src/rust/engine/process_execution/src/immutable_inputs.rs

```diff
@@ -39,28 +43,47 @@ impl ImmutableInputs {
   /// Returns an absolute Path to immutably consume the given Digest from.
   async fn path(&self, digest: Digest) -> Result<PathBuf, String> {
     let cell = self.contents.lock().entry(digest).or_default().clone();
-    let value: Result<_, String> = cell
+    let value = cell
       .get_or_try_init(async {
-        let digest_str = digest.hash.to_hex();
-
-        let path = self.workdir.path().join(digest_str);
-        if let Ok(meta) = tokio::fs::metadata(&path).await {
-          // TODO: If this error triggers, it indicates that we have previously checked out this
-          // directory, either due to a race condition, or due to a previous failure to
-          // materialize. See https://github.com/pantsbuild/pants/issues/13899
-          return Err(format!(
-            "Destination for immutable digest already exists: {:?}",
-            meta
-          ));
-        }
+        let chroot = TempDir::new_in(self.workdir.path()).map_err(|e| {
+          format!(
+            "Failed to create a temporary directory for materialization of immutable input \
+            digest {:?}: {}",
+            digest, e
+          )
+        })?;
         self
           .store
-          .materialize_directory(path.clone(), digest, Permissions::ReadOnly)
+          .materialize_directory(chroot.path().to_path_buf(), digest, Permissions::ReadOnly)
+          .await?;
+        let materialized_chroot = chroot.into_path();
+        let digest_str = digest.hash.to_hex();
+        let path = self.workdir.path().join(&digest_str);
+        tokio::fs::rename(&materialized_chroot, &path)
```

Comment:
So it turns out you cannot rename a 555 dir under macOS; thus the perm denied on macOS only in the tests right now. Who knew? I <3 macOS.